README.rst
nullunit
========

This is the agent that runs tasks for MCP.


debian-common/changelog
nullunit (0.45-1) DIST; urgency=medium

  * fixes for centoses

 -- Peter Howe <pnhowe@gmail.com>  Thu, 29 Oct 2020 10:44:12 -0600

nullunit (0.44-1) DIST; urgency=medium

  * Retry "apt-get update"s

 -- Peter Howe <pnhowe@gmail.com>  Fri, 23 Oct 2020 07:41:03 -0600

nullunit (0.43-1) DIST; urgency=medium

  * Updates for MCP 0.23

 -- Peter Howe <pnhowe@gmail.com>  Mon, 03 Aug 2020 06:54:02 -0600

nullunit (0.42-1) DIST; urgency=medium

  * nullunitInterface fixes for mcp's auto-build change

 -- Peter Howe <peter.howe@virtustream.com>  Mon, 09 Sep 2019 12:03:41 -0600

nullunit (0.41-1) DIST; urgency=medium

  * nullunitInterface fixes

 -- Peter Howe <peter.howe@virtustream.com>  Wed, 21 Aug 2019 20:48:43 -0600

nullunit (0.40-1) DIST; urgency=medium

  * Confluence update attachment fix

 -- Peter Howe <peter.howe@virtustream.com>  Thu, 08 Aug 2019 22:29:23 -0600

nullunit (0.39-1) DIST; urgency=medium

  * fix

 -- Peter Howe <peter.howe@virtustream.com>  Thu, 08 Aug 2019 08:13:02 -0600

nullunit (0.38-1) DIST; urgency=medium

  * confluence fixes

 -- Peter Howe <peter.howe@virtustream.com>  Wed, 07 Aug 2019 23:27:51 -0600

nullunit (0.37-1) DIST; urgency=medium

  * goes with MCP 0.20
  * new config file location
  * documentation updates

 -- Peter Howe <peter.howe@virtustream.com>  Tue, 06 Aug 2019 16:14:02 -0600

nullunit (0.36-1) DIST; urgency=medium

  * Adding Doc support

 -- Peter Howe <peter.howe@virtustream.com>  Fri, 02 Aug 2019 18:09:36 -0600

nullunit (0.35-1) DIST; urgency=medium

  * goes with MCP 0.19

 -- Peter Howe <peter.howe@virtustream.com>  Thu, 18 Jul 2019 12:43:28 -0600

nullunit (0.34-1) DIST; urgency=medium

  * bug fix for targets without requirements
  * reworking file uploading to packrat handeling

 -- Peter Howe <peter.howe@virtustream.com>  Tue, 02 Jul 2019 11:14:44 -0600

nullunit (0.33-1) DIST; urgency=medium

  * MCP now detects when the job is done, remove extra target detection
    and manual signaling

 -- Peter Howe <peter.howe@virtustream.com>  Mon, 01 Jul 2019 14:37:29 -0600

nullunit (0.32-1) DIST; urgency=medium

  * filepath with distroversion and type reworked

 -- Peter Howe <peter.howe@virtustream.com>  Fri, 17 May 2019 22:26:51 -0600

nullunit (0.31-1) DIST; urgency=medium

  * packrat psk renamed to password

 -- Peter Howe <peter.howe@virtustream.com>  Fri, 17 May 2019 09:35:06 -0600

nullunit (0.30-1) DIST; urgency=medium

  * added build_name
  * updated for new packrat

 -- Peter Howe <peter.howe@virtustream.com>  Tue, 07 May 2019 13:19:35 -0600

nullunit (0.29-1) DIST; urgency=medium

  * Detect the cpu core count and pass that as -j for the target build

 -- Peter Howe <peter.howe@virtustream.com>  Thu, 13 Dec 2018 07:20:52 -0700

nullunit (0.28-1) DIST; urgency=medium

  * Bumped added MCP version 0.10
  * added centos-7

 -- Peter Howe <peter.howe@virtustream.com>  Fri, 07 Dec 2018 14:30:49 -0700

nullunit (0.27-5) DIST; urgency=medium

  * updated nullunitInterface for the new mcp/contractor interface

 -- Peter Howe <peter.howe@virtustream.com>  Mon, 27 Aug 2018 08:12:23 -0600

nullunit (0.26-5) DIST; urgency=medium

  * combining test/lint
  * flag to enable/disable uploading packages

 -- Peter Howe <peter.howe@virtustream.com>  Mon, 06 Aug 2018 16:02:32 -0600

nullunit (0.25-3) DIST; urgency=medium

  * Test score extraction

 -- Peter Howe <peter.howe@virtustream.com>  Fri, 20 Jul 2018 12:42:18 -0600

nullunit (0.24-1) DIST; urgency=medium

  * Packaging updates
  * renamed docs to doc

 -- Peter Howe <peter.howe@virtustream.com>  Mon, 16 Jul 2018 23:20:38 -0600

nullunit (0.23-3) DIST; urgency=medium

  * http(s)_proxy env vars

 -- Peter Howe <peter.howe@virtustream.com>  Fri, 13 Jul 2018 08:55:57 -0600

nullunit (0.22-2) DIST; urgency=medium

  * Packaging tweeks

 -- Peter Howe <peter.howe@virtustream.com>  Thu, 12 Jul 2018 12:21:41 -0600

nullunit (0.22-1) DIST; urgency=medium

  * Ported to Python3
  * Updated to work with new MCP that uses Contractor
  * Updated to to packrat api 1.5

 -- Peter Howe <peter.howe@virtustream.com>  Sat, 07 Jul 2018 22:09:34 -0600

nullunit (0.21-1) DIST; urgency=medium

  * updates for docs target

 -- Peter Howe <peter@blueboy>  Tue, 25 Oct 2016 23:57:58 -0600

nullunit (0.20-1) DIST; urgency=medium

  * Changes to Accomadate new Packrat

 -- Peter Howe <peter@blueboy>  Thu, 07 Jul 2016 14:18:02 -0600

nullunit (0.19-3) DIST; urgency=medium

  * lint and test score returnd to MCP
  * package list returnd to MCP

 -- Peter Howe <peter@blueboy>  Fri, 15 Apr 2016 22:45:00 -0600

nullunit (0.18-7) DIST; urgency=medium

  * change workd dir to help out go builds
  * saving more output

 -- Peter Howe <peter@blueboy>  Thu, 10 Mar 2016 10:09:02 -0700

nullunit (0.17-3) DIST; urgency=medium

  * run <target>-setup for all targets,
    not just package ones

 -- Peter Howe <peter@blueboy>  Tue, 02 Feb 2016 15:17:33 -0700

nullunit (0.16-2) DIST; urgency=medium

  * reboot after getting config to make sure
    system is cleanly loaded as it's new id

 -- Peter Howe <peter@blueboy>  Mon, 25 Jan 2016 15:05:16 -0700

nullunit (0.15-6) DIST; urgency=medium

  * support for network resources
  * support for pushing config values
  * support for reading config values
  * nullunit now runs all at once

 -- Peter Howe <peter@blueboy>  Mon, 18 Jan 2016 10:31:41 -0700

nullunit (0.14-2) DIST; urgency=medium

  * to a repo update before installing requirements

 -- Peter Howe <peter@blueboy>  Tue, 12 Jan 2016 16:41:39 -0700

nullunit (0.13-1) DIST; urgency=medium

  * Preallocation support

 -- Peter Howe <peter@blueboy>  Tue, 12 Jan 2016 08:26:23 -0700

nullunit (0.12-1) DIST; urgency=medium

  * handeling multiple package files per line

 -- Peter Howe <peter@blueboy>  Wed, 02 Sep 2015 08:18:00 -0600

nullunit (0.11-1) DIST; urgency=medium

  * nullunitMastersync tweekes
  * bug fixes

 -- Peter Howe <peter@blueboy>  Wed, 10 Jun 2015 16:30:23 -0600

nullunit (0.10-3) DIST; urgency=medium


nullunit/MCP.py
import logging
import time
import random
import math

from cinp import client

MCP_API_VERSIONS = ( '0.9', '0.10' )

# TODO: move backoff delay and retries to cinp client
DELAY_MULTIPLIER = 15
# delay of 15 results in a delay of:
# min delay = 0, 10, 16, 20, 24, 26, 29, 31, 32, 34, 35, 37, 38, 39, 40 ....
# max delay = 0, 20, 32, 40, 48, 52, 58, 62, 64, 68, 70, 74, 76, 78, 80 .....


def _backOffDelay( count ):
  if count < 1:  # math.log dosen't do so well below 1
    count = 1

  if count > 20:  # really don't need to get any more delaied than this
    count = 20

  factor = int( DELAY_MULTIPLIER * math.log( count ) )
  delay = int( factor + ( random.random() * factor ) )
  logging.debug( 'MCP: sleeping for "{0}"'.format( delay ) )
  time.sleep( delay )


class MCP( object ):
  def __init__( self, host, proxy, job_id, instance_id, cookie ):
    self.cinp = client.CInP( host, '/api/v1/', proxy )
    self.job_id = job_id
    self.instance_id = instance_id
    self.cookie = cookie

    count = 0
    while True:
      count += 1
      try:
        root = self.cinp.describe( '/api/v1/' )
        break
      except ( client.Timeout, client.ResponseError ):
        _backOffDelay( count )
        logging.warn( 'MCP: getRequest: retry {0}'.format( count ) )

    if root[ 'api-version' ] not in MCP_API_VERSIONS:
      raise Exception( 'Expected API version (one of) "{0}" found "{1}"'.format( MCP_API_VERSIONS, root[ 'api-version' ] ) )

  def contractorInfo( self ):
    logging.info( 'MCP: Get Contractor Info' )
    info = self.cinp.call( '/api/v1/config(getContractorInfo)', {} )

    return { 'host': info[ 'host' ], 'proxy': self.cinp.proxy }

  def signalJobRan( self ):
    logging.info( 'MCP: Signal Job Ran' )
    self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(jobRan)'.format( self.instance_id ), { 'cookie': self.cookie } )

  def sendMessage( self, message ):
    logging.info( 'MCP: Message "{0}"'.format( message ) )
    self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(setMessage)'.format( self.instance_id ), { 'cookie': self.cookie, 'message': message } )

  def setSuccess( self, success ):
    logging.info( 'MCP: Success "{0}"'.format( success ) )
    self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(setSuccess)'.format( self.instance_id ), { 'cookie': self.cookie, 'success': success } )

  def setResults( self, target, results ):
    if results is not None:
      logging.info( 'MCP: Results "{0}"'.format( results[ -100: ].strip() ) )
    else:
      logging.info( 'MCP: Results <empty>' )

    self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(setResults)'.format( self.instance_id ), { 'cookie': self.cookie, 'target': target, 'results': results } )

  def setScore( self, target, score ):
    if score is not None:
      logging.info( 'MCP: Score "{0}"'.format( score ) )
    else:
      logging.info( 'MCP: Score <undefined>' )

    self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(setScore)'.format( self.instance_id ), { 'cookie': self.cookie, 'target': target, 'score': score } )

  def uploadedPackages( self, package_file_map ):
    if not package_file_map:
      return

    self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(addPackageFiles)'.format( self.instance_id ), { 'cookie': self.cookie, 'package_file_map': package_file_map } )

  def getInstanceState( self, name=None ):
    logging.info( 'MCP: Instance State for "{0}"'.format( name ) )
    args = {}
    if name is not None:
      args[ 'name' ] = name

    # json encoding turns the numeric dict keys into strings, this will undo that  # TODO:  this is fixed in CInP now??
    result = {}
    state_map = self.cinp.call( '/api/v1/Processor/BuildJob:{0}:(getInstanceState)'.format( self.job_id ), args )
    if name is None:
      for name in state_map:
        result[ name ] = {}
        for index, state in state_map[ name ].items():
          result[ name ][ int( index ) ] = state
    else:
      for index, state in state_map.items():
        result[ int( index ) ] = state

    return result

  def getInstanceDetail( self, name=None ):
    logging.info( 'MCP: Instance Detail for "{0}"'.format( name ) )
    args = {}
    if name is not None:
      args[ 'name' ] = name

    # json encoding turns the numeric dict keys into strings, this will undo that
    result = {}
    detail_map = self.cinp.call( '/api/v1/Processor/BuildJob:{0}:(getInstanceDetail)'.format( self.job_id ), args )
    if name is None:
      for name in detail_map:
        result[ name ] = {}
        for index, detail in detail_map[ name ].items():
          result[ name ][ int( index ) ] = detail
    else:
      for index, detail in detail_map.items():
        result[ int( index ) ] = detail

    return result

  def updateValueMap( self, value_map  ):
    logging.info( 'MCP: Setting Value "{0}"'.format( value_map ) )

    self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(updateValueMap)'.format( self.instance_id ), { 'cookie': self.cookie, 'value_map': value_map } )

    return True

  def getValueMap( self, name=None ):
    logging.info( 'MCP: Getting Value Map' )

    return self.cinp.call( '/api/v1/Processor/BuildJobResourceInstance:{0}:(getValueMap)'.format( self.instance_id ), { 'cookie': self.cookie } )
